File: wqpu_next/scheduler.py
"""Deterministic WQPU global price controller and distributed scheduler."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List, Sequence

from .protocol import BPS, GlobalPriceState, ProviderRecord

TARGET_UTILIZATION_BPS = 7000
MAX_PRICE_MOVE_BPS = 500
MIN_PRICE_PER_MILLION = 1
DEFAULT_MAX_WORKERS = 8
MODEL_HEADROOM_BPS = 1200  # keep 12% advertised memory unused


@dataclass(frozen=True)
class WorkerAllocation:
    wallet: str
    peer_id: str
    endpoint: str
    assigned_model_bytes: int
    utilization_bps: int
    free_units: int


@dataclass(frozen=True)
class SchedulePlan:
    model_hash: str
    total_model_bytes: int
    allocations: List[WorkerAllocation]

    @property
    def assigned_model_bytes(self) -> int:
        return sum(x.assigned_model_bytes for x in self.allocations)


def aggregate_price_state(
    providers: Iterable[ProviderRecord],
    previous_price_per_million: int,
    next_epoch: int,
) -> GlobalPriceState:
    """Calculate the one network-wide compute price for the next epoch.

    The controller is deliberately simple and consensus-friendly:
    - all arithmetic is integer-only;
    - target utilization is 70%;
    - price can move by at most 5% per epoch;
    - the same provider snapshot always produces the same answer.
    """
    if previous_price_per_million <= 0:
        raise ValueError("previous price must be positive")
    if next_epoch < 0:
        raise ValueError("epoch must be non-negative")

    items = list(providers)
    capacity = sum(max(0, p.capacity_units) for p in items)
    busy = sum(min(max(0, p.busy_units), max(0, p.capacity_units)) for p in items)

    if capacity <= 0:
        utilization_bps = BPS
    else:
        utilization_bps = min(BPS, (busy * BPS) // capacity)

    deviation = utilization_bps - TARGET_UTILIZATION_BPS
    # Quarter-strength response prevents oscillation; hard-bound to +/- 5%.
    move_bps = deviation // 4
    move_bps = max(-MAX_PRICE_MOVE_BPS, min(MAX_PRICE_MOVE_BPS, move_bps))

    numerator = previous_price_per_million * (BPS + move_bps)
    price = max(MIN_PRICE_PER_MILLION, numerator // BPS)

    return GlobalPriceState(
        epoch=next_epoch,
        price_per_million_units=price,
        aggregate_capacity_units=capacity,
        aggregate_busy_units=busy,
    )


def _usable_memory_bytes(provider: ProviderRecord) -> int:
    headroom = (provider.free_memory_bytes * MODEL_HEADROOM_BPS) // BPS
    return max(0, provider.free_memory_bytes - headroom)


def compatible_providers(
    providers: Iterable[ProviderRecord],
    model_hash: str,
    at_height: int,
) -> List[ProviderRecord]:
    out: List[ProviderRecord] = []
    for provider in providers:
        try:
            provider.validate(at_height=at_height)
        except ValueError:
            continue
        if model_hash not in provider.model_hashes:
            continue
        if provider.free_units <= 0:
            continue
        if _usable_memory_bytes(provider) <= 0:
            continue
        out.append(provider)
    return out


def select_least_busy(
    providers: Iterable[ProviderRecord],
    model_hash: str,
    model_bytes: int,
    at_height: int,
    max_workers: int = DEFAULT_MAX_WORKERS,
) -> SchedulePlan:
    """Select enough least-busy compatible peers to hold one model collectively.

    No worker is assumed to contain the whole model. The scheduler fills model
    bytes across workers, starting with the lowest utilization. Free capacity and
    memory are deterministic tie-breakers; wallet/peer IDs make the final order
    stable across coordinators reading the same chain snapshot.
    """
    if not model_hash:
        raise ValueError("model_hash must be non-empty")
    if model_bytes <= 0:
        raise ValueError("model_bytes must be positive")
    if max_workers <= 0:
        raise ValueError("max_workers must be positive")

    candidates = compatible_providers(providers, model_hash, at_height)
    candidates.sort(
        key=lambda p: (
            p.utilization_bps,
            -p.free_units,
            -_usable_memory_bytes(p),
            p.wallet,
            p.peer_id,
        )
    )

    allocations: List[WorkerAllocation] = []
    remaining = model_bytes

    for provider in candidates[:max_workers]:
        if remaining <= 0:
            break
        usable_memory = _usable_memory_bytes(provider)
        assigned = min(remaining, usable_memory)
        if assigned <= 0:
            continue
        allocations.append(
            WorkerAllocation(
                wallet=provider.wallet,
                peer_id=provider.peer_id,
                endpoint=provider.endpoints[0],
                assigned_model_bytes=assigned,
                utilization_bps=provider.utilization_bps,
                free_units=provider.free_units,
            )
        )
        remaining -= assigned

    if remaining > 0:
        raise RuntimeError(
            "insufficient compatible free memory: missing {} bytes".format(remaining)
        )

    return SchedulePlan(
        model_hash=model_hash,
        total_model_bytes=model_bytes,
        allocations=allocations,
    )


def charge_for_units(price_per_million_units: int, compute_units: int) -> int:
    """Deterministic integer charge, rounded up so tiny jobs are not free."""
    if price_per_million_units <= 0:
        raise ValueError("price must be positive")
    if compute_units < 0:
        raise ValueError("compute units must be non-negative")
    if compute_units == 0:
        return 0
    numerator = price_per_million_units * compute_units
    return (numerator + 999_999) // 1_000_000

File: chain/x/wqpu/kernel/scheduler.go
package kernel

import (
	"errors"
	"math/big"
	"sort"
)

func mulDivFloor(a, b, d uint64) (uint64, error) {
	if d == 0 {
		return 0, errors.New("division by zero")
	}
	n := new(big.Int).Mul(new(big.Int).SetUint64(a), new(big.Int).SetUint64(b))
	n.Quo(n, new(big.Int).SetUint64(d))
	if !n.IsUint64() {
		return 0, errors.New("uint64 overflow")
	}
	return n.Uint64(), nil
}

func mulDivCeil(a, b, d uint64) (uint64, error) {
	if d == 0 {
		return 0, errors.New("division by zero")
	}
	n := new(big.Int).Mul(new(big.Int).SetUint64(a), new(big.Int).SetUint64(b))
	q, r := new(big.Int), new(big.Int)
	q.QuoRem(n, new(big.Int).SetUint64(d), r)
	if r.Sign() != 0 {
		q.Add(q, big.NewInt(1))
	}
	if !q.IsUint64() {
		return 0, errors.New("uint64 overflow")
	}
	return q.Uint64(), nil
}

// floorDivSigned matches Python's // semantics for consensus math.
func floorDivSigned(n, d int64) int64 {
	if d <= 0 {
		panic("floorDivSigned requires a positive divisor")
	}
	q := n / d
	if n < 0 && n%d != 0 {
		q--
	}
	return q
}

func AggregatePriceState(capacity, reserved, previousPrice, nextEpoch uint64) (GlobalPriceState, error) {
	if previousPrice == 0 {
		return GlobalPriceState{}, errors.New("previous price must be positive")
	}

	busy := reserved
	if capacity == 0 {
		busy = 0
	} else if busy > capacity {
		busy = capacity
	}

	utilization := BasisPoints
	if capacity > 0 {
		var err error
		utilization, err = mulDivFloor(busy, BasisPoints, capacity)
		if err != nil {
			return GlobalPriceState{}, err
		}
		if utilization > BasisPoints {
			utilization = BasisPoints
		}
	}

	deviation := int64(utilization) - int64(TargetUtilizationBPS)
	move := floorDivSigned(deviation, 4)
	if move > int64(MaxPriceMoveBPS) {
		move = int64(MaxPriceMoveBPS)
	}
	if move < -int64(MaxPriceMoveBPS) {
		move = -int64(MaxPriceMoveBPS)
	}

	factor := int64(BasisPoints) + move
	if factor <= 0 {
		return GlobalPriceState{}, errors.New("invalid price factor")
	}
	price, err := mulDivFloor(previousPrice, uint64(factor), BasisPoints)
	if err != nil {
		return GlobalPriceState{}, err
	}
	if price < MinPricePerMillion {
		price = MinPricePerMillion
	}

	return GlobalPriceState{
		Epoch:                  nextEpoch,
		PricePerMillionUnits:   price,
		AggregateCapacityUnits: capacity,
		AggregateBusyUnits:     busy,
	}, nil
}

func ChargeForUnits(pricePerMillion, computeUnits uint64) (uint64, error) {
	if pricePerMillion == 0 {
		return 0, errors.New("price must be positive")
	}
	if computeUnits == 0 {
		return 0, nil
	}
	return mulDivCeil(pricePerMillion, computeUnits, 1_000_000)
}

func containsModel(models []string, wanted string) bool {
	for _, model := range models {
		if model == wanted {
			return true
		}
	}
	return false
}

func effectiveBusy(p Provider, reserved map[string]uint64) uint64 {
	busy := p.BusyUnits
	if r := reserved[p.PeerID]; r > busy {
		busy = r
	}
	if busy > p.CapacityUnits {
		busy = p.CapacityUnits
	}
	return busy
}

func effectiveFree(p Provider, reserved map[string]uint64) uint64 {
	busy := effectiveBusy(p, reserved)
	if busy >= p.CapacityUnits {
		return 0
	}
	return p.CapacityUnits - busy
}

func effectiveUtilization(p Provider, reserved map[string]uint64) uint64 {
	if p.CapacityUnits == 0 {
		return BasisPoints
	}
	v, err := mulDivFloor(effectiveBusy(p, reserved), BasisPoints, p.CapacityUnits)
	if err != nil || v > BasisPoints {
		return BasisPoints
	}
	return v
}

func usableMemory(p Provider) uint64 {
	headroom, err := mulDivFloor(p.FreeMemoryBytes, ModelHeadroomBPS, BasisPoints)
	if err != nil || headroom >= p.FreeMemoryBytes {
		return 0
	}
	return p.FreeMemoryBytes - headroom
}

func SelectLeastBusy(
	providers []Provider,
	modelHash string,
	modelBytes uint64,
	atHeight uint64,
	maxWorkers int,
	reservedByPeer map[string]uint64,
) (SchedulePlan, error) {
	if modelHash == "" {
		return SchedulePlan{}, errors.New("model hash must be non-empty")
	}
	if modelBytes == 0 {
		return SchedulePlan{}, errors.New("model size must be positive")
	}
	if maxWorkers <= 0 {
		return SchedulePlan{}, errors.New("max workers must be positive")
	}
	if reservedByPeer == nil {
		reservedByPeer = map[string]uint64{}
	}

	candidates := make([]Provider, 0, len(providers))
	for _, p := range providers {
		if p.Validate(atHeight) != nil {
			continue
		}
		if !containsModel(p.ModelHashes, modelHash) {
			continue
		}
		if effectiveFree(p, reservedByPeer) == 0 || usableMemory(p) == 0 {
			continue
		}
		candidates = append(candidates, p)
	}

	sort.SliceStable(candidates, func(i, j int) bool {
		a, b := candidates[i], candidates[j]
		au, bu := effectiveUtilization(a, reservedByPeer), effectiveUtilization(b, reservedByPeer)
		if au != bu {
			return au < bu
		}
		af, bf := effectiveFree(a, reservedByPeer), effectiveFree(b, reservedByPeer)
		if af != bf {
			return af > bf
		}
		am, bm := usableMemory(a), usableMemory(b)
		if am != bm {
			return am > bm
		}
		if a.Wallet != b.Wallet {
			return a.Wallet < b.Wallet
		}
		return a.PeerID < b.PeerID
	})

	if len(candidates) > maxWorkers {
		candidates = candidates[:maxWorkers]
	}

	remaining := modelBytes
	allocations := make([]WorkerAllocation, 0, len(candidates))
	for _, p := range candidates {
		if remaining == 0 {
			break
		}
		assigned := usableMemory(p)
		if assigned > remaining {
			assigned = remaining
		}
		if assigned == 0 {
			continue
		}
		allocations = append(allocations, WorkerAllocation{
			Wallet:             p.Wallet,
			PeerID:             p.PeerID,
			Endpoint:           p.Endpoints[0],
			AssignedModelBytes: assigned,
			UtilizationBPS:     effectiveUtilization(p, reservedByPeer),
			FreeUnits:          effectiveFree(p, reservedByPeer),
		})
		remaining -= assigned
	}

	if remaining != 0 {
		return SchedulePlan{}, errors.New("insufficient compatible free memory")
	}

	return SchedulePlan{
		ModelHash:       modelHash,
		TotalModelBytes: modelBytes,
		Allocations:     allocations,
	}, nil
}
